src/pages/MainPage/useProductList.ts

```diff
@@ -0,0 +1,47 @@
+import { useEffect, useMemo } from 'react';
+import { useDispatch, useSelector } from 'react-redux';
+import { actions } from '../../redux/actions';
+import { StoreState, Product } from '../../types';
+
+type SelectedState = Pick<
+  StoreState['productsState'],
+  'isLoading' | 'productList' | 'error'
+>;
+
+const useProductList = () => {
+  const dispatch = useDispatch();
+  const { isLoading, productList, error } = useSelector<
+    StoreState,
+    SelectedState
+  >(({ productsState }) => ({
+    isLoading: productsState.isLoading,
+    productList: productsState.productList,
+    error: productsState.error,
+  }));
+
+  useEffect(() => {
+    dispatch(actions.getProductList());
+  }, [dispatch]);
+
+  const sortedProductList = useMemo(() => {
+    const { availableList, outOfStockList } = productList.reduce(
+      (next, product) => {
+        const isAvailable = product.stock > 0;
+
+        (isAvailable ? next.availableList : next.outOfStockList).push(product);
+
+        return next;
+      },
+      {
+        availableList: [] as Array<Product>,
+        outOfStockList: [] as Array<Product>,
+      }
+    );
```

Comment:
```suggestion
const availableList = productList.filter((product) => product.stock > 0);
const outOfStockList = productList.filter((product) => product.stock <= 0);
```

코드를 가독성 좋게 작성해보는건 어떨까요?